addons/mail/models/models.py

```diff
@@ -185,17 +185,22 @@ def _notify_get_reply_to_formatted_email(self, record_email, record_name):
 
     def _alias_get_error_message(self, message, message_dict, alias):
         """ Generic method that takes a record not necessarily inheriting from
-        mail.alias.mixin. """
+        mail.alias.mixin.
```

Comment:
I wonder if it would not be better to have two or three methods (each with their clear input / output and purpose), and use error keys. Something like
```
def _alias_get_error_key(...):
    ... # blah
    return 'config_follower_no_record' / 'config_follower_no_field' /
          'error_follower_not_following' / 'error_partners_no_partner'

def _alias_get_error_message(key):
    if key == 'config_follower_no_record':
        return _('incorrectly configured alias (unknown reference record)'
    return _('unknown error')

def _alias_is_error_from_config(key):
    """ True if the error seems linked to alias configuration, not sender
    identity. """"
    return key in ['config_follower_no_record', 'config_follower_no_field']
```

It makes code a bit more verbose, but I find it a bit more easier to understand and follow. Moreover it allows to play with keys that are easy to grep. What do you think ?

cc @std-odoo

Reply:
Yes the returned tuple is not the best. While changing it, I feel it complicate to link all parts together, if you add a key in _alias_get_error_key, you have to not forget to add it in the 2 other functions. And if you want to add details in the error, you don’t have it anymore in _alias_get_error_message.

So I have another proposition which is just instead of returning a tuple, returning a dataclass (supported in python 3.7, minimal required version of odoo) but to ease search we can add an error code as well as suggested (equality tested only on the code):
```python
@dataclass(frozen=True)
class AliasError:
    code: str
    message: str = field(default='', compare=False)
    is_config_error: bool = field(default=False, compare=False)

# ...
def _alias_get_error(self, message, message_dict, alias):
    if alias.alias_contact == 'followers':
        if not self.ids:
            return AliasError('config_follower_no_record',
                              _('incorrectly configured alias (unknown reference record)'),
                              True)
    # ...
    return False
```
which displays as: 
alias_error = AliasError(code='config_follower_no_record', message='incorrectly configured alias (unknown reference record)', is_config_error=True)
and still allow comparison:
alias_error == AliasError('config_follower_no_record') → True

What do you think ?

Reply:
I don't know anything about dataclasses :D so curious to look at what it will give :)

Reply:
I have just updated the code with the dataclass. It is new in python 3.7 and a bit like the namedtuple but more flexible.